sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3819,8 +3841,29 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
 
   ZE_CALL(zeKernelSetGroupSize, (Kernel->ZeKernel, WG[0], WG[1], WG[2]));
 
+  pi_platform Plt = Kernel->Program->Context->Devices[0]->Platform;
   // Lock automatically releases when this goes out of scope.
-  std::lock_guard<std::mutex> lock(Queue->PiQueueMutex);
+  std::lock(Queue->PiQueueMutex, Plt->MemAllocsMutex);
+  std::lock_guard<std::mutex> QueueLock(Queue->PiQueueMutex, std::adopt_lock);
+  std::lock_guard<std::mutex> MemAllocsLock(Plt->MemAllocsMutex,
+                                            std::adopt_lock);
+
+  // We are going to submit a kernel for execution. If indirect access flag is
+  // set for a kernel then we need to make a snapshot of existing memory
+  // allocations. We need to lock the mutex guarding the list of memory
+  // allocations in the platform to prevent creation of new memory alocations
+  // before we submit the kernel for execution. Mutex guaring the list of
+  // referenced allocations in the kernel must be locked as well because kernel
+  // can be submitted multiple times from different threads.
+  // Note. Currently indirect access flag is set for all kernels to be
+  // conservative, that's why no need to check for a flag.
+  for (auto &MemAlloc : Plt->MemAllocs) {
+    Kernel->MemAllocs.push_back(&MemAlloc);
+    // Kernel is referencing this memory allocation from now.
+    MemAlloc.RefCount++;
```

Comment:
I think this would have bad interaction with commands batching, i.e. new memory can be allocated before the current batch command-list is actually closed and submitted.

Reply:
Thanks for catching this. Moved capturing to the place where command-list is actually closed and submitted.